junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ResourceLock.java

```diff
@@ -43,12 +46,47 @@ default void close() {
 		release();
 	}
 
+	/**
+	 * {@return the exclusive resources this lock represents}
+	 */
+	List<ExclusiveResource> getResources();
+
+	/**
+	 * {@return whether this lock requires exclusiveness}
+	 */
+	boolean isExclusive();
+
 	/**
 	 * {@return whether the given lock is compatible with this lock}
 	 * @param other the other lock to check for compatibility
 	 */
 	default boolean isCompatible(ResourceLock other) {
-		return this instanceof NopLock || other instanceof NopLock;
-	}
 
+		List<ExclusiveResource> ownResources = this.getResources();
+		List<ExclusiveResource> otherResources = other.getResources();
+
+		if (ownResources.isEmpty() || otherResources.isEmpty()) {
+			return true;
+		}
+
+		// Whenever there's a READ_WRITE lock, it's incompatible with any other lock
+		// because we guarantee that all children will have exclusive access to the
+		// resource in question. In practice, whenever a READ_WRITE lock is present,
+		// NodeTreeWalker will force all children to run in the same thread so that
+		// it should never attempt to steal work from another thread, and we shouldn't
+		// actually reach this point.
+		boolean isGlobalReadLock = ownResources.size() == 1
+				&& ExclusiveResource.GLOBAL_READ.equals(ownResources.get(0));
+		if ((!isGlobalReadLock && other.isExclusive()) || this.isExclusive()) {
```

Comment:
Why not just
```suggestion
		if (other.isExclusive() || this.isExclusive()) {
```

Reply:
Because we need to make the global read lock (which is always on direct children of the engine node) compatible with the first write lock (which may be on a test method). 

This was already the case before this set of changes:
https://github.com/junit-team/junit5/blob/81d347986208573a654196dfcb3e2ed09ab493a5/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/SingleLock.java#L63-L72

Reply:
Done in ec118028e190703a0f59cf77cfc9c1ae52b23da1.